[docs] Add module development documentation (#7779)
deckhouse/deckhouse: docs/documentation/pages/modules-docs/README_RU.md

---
title: "Модули Deckhouse Kubernetes Platform"
permalink: ru/modules-docs/
lang: ru
---

<p align="center">
<img src="../images/modules-docs/modules_logo.png" alt="deckhouse modules logo" />
</p>

> ## Введение

Reviewer: ```suggestion

```

* 📚 Прочитайте документацию операторов о концепции хуков, например, [что такое конфигурация хука и какие функции она предоставляет](https://flant.github.io/shell-operator/HOOKS.html#hook-configuration). При помощи конфигурации настраиваются данные, которые будут доступны из хука.
* Ознакомьтесь с информацией о [биндингах](https://flant.github.io/addon-operator/HOOKS.html#bindings). Биндинги - это события, при которых срабатывает хук. Биндинги указываются в конфигурации хука. Хук может сработать не только из-за событий в Kubernetes, а также, например, по расписанию или перед запуском модуля.
> Хук позволяет сохранять значения в памяти и использовать их позже при рендеринге шаблонов Helm. Об этой особенности и цикле работы модуля, рекомендуем прочитать в [документации Hooks and Helm values](https://flant.github.io/addon-operator/OVERVIEW.html#hooks-and-helm-values).
* Ознакомьтесь с [концепцией снепшотов](https://flant.github.io/shell-operator/HOOKS.html#snapshots). С помощью снепшотов можно игнорировать отдельные события и реализовать шаблон reconciliation loop, при котором состояние из снепшота приводится к состоянию модуля.

>  > Этим способом в Deckhouse Kubernetes Platform реализовано 100% поддержка хуков во внутренних модулях.

Reviewer: ```suggestion
 > Этим способом в Deckhouse Kubernetes Platform реализована 100% поддержка хуков во внутренних модулях.
```

 * Кроме того, хуки можно использовать вместо `prometheus exporter`. Хуки могут предоставлять метрики, которые Deckhouse будет экспортировать. Ознакомьтесь с информацией [о метриках](https://flant.github.io/addon-operator/metrics/METRICS_FROM_HOOKS.html#custom-metrics).

> * 🎬 В [видео 2019](https://www.youtube.com/watch?v=1_55KPHjVTU) года @andrey.polovov подробно рассказал о том, что такое shell-operator и addon-operator.

Reviewer: ```suggestion
* 🎬 В [видео 2019](https://www.youtube.com/watch?v=1_55KPHjVTU) года Андрей Половов подробно рассказал о том, что такое shell-operator и addon-operator.
```

---

> title: "Модули Deckhouse"

Reviewer 2: ```suggestion
title: "Модули Deckhouse Kubernetes Platform"
```

permalink: ru/modules-docs/
lang: ru
---

<p align="center">
<img src="../images/modules-docs/modules_logo.png" alt="deckhouse modules logo" />
</p>

## Введение

> В этом репозитории вы найдете документацию для создания своего модуля Deckhouse.

Reviewer 2: ```suggestion
В этом репозитории представлена документация для создания собственного модуля Deckhouse.
```

> ## Для ознакомления до начала чтения

Reviewer 2: ```suggestion
## Перед чтением, ознакомьтесь с этой информацией
```

> Принципы, по которым работают модули Deckhouse, можно понять ознакомившись с [addon-operator](https://github.com/flant/addon-operator) и [shell-operator](https://github.com/flant/addon-operator).
> 
> * 📚 Про концепцию хуков можно почерпнуть из документации операторов
>   * [Что такое конфигурация хука и какие есть опции](https://flant.github.io/shell-operator/HOOKS.html#hook-configuration). При помощи конфигурации мы настраиваем как будет выглядеть данные, которые будут доступны из хука.
>   * Отдельного внимания достойны [биндинги](https://flant.github.io/addon-operator/HOOKS.html#bindings) -  события, при которых будет срабатывать хук. Биндинги указываются в конфигурации хука. Хук может срабатывать не тольк из-за событий в Kubernetes, но и, например, по расписанию, или каждый раз перед запуском модуля.
>   * Хук может сохранять значения в память и использовать потом при рендере шаблонов Helm. Подробнее про эту особенность и цикл работы модуля можно прочитать в [этой документации](https://flant.github.io/addon-operator/OVERVIEW.html#hooks-and-helm-values).
>   * Обязательного ознакомления требует [концепция снепшотов](https://flant.github.io/shell-operator/HOOKS.html#snapshots). При помощи снепшотов можно перестать реагировать на отдельные события и реализовать паттерн reconciliation loop, приводя состояние из снепшота к состоянию модуля.
>       > Этим способом в Deckhouse реализовано 100% хуков во внутренних модулях.
>   * Можно использовать хук как замену `prometheus exporter`. Хуки могут возвращать метрики, которые будет экспортировать Deckhouse. Подробнее о метриках можно [прочитать тут](https://flant.github.io/addon-operator/metrics/METRICS_FROM_HOOKS.html#custom-metrics).
> * 🎬 В [видео 2019](https://www.youtube.com/watch?v=1_55KPHjVTU) года @andrey.polovov подробно рассказал о том, что такое shell-operator и addon-operator.
> * 🎬 [Более подробное видео](https://www.youtube.com/watch?v=we0s4ETUBLc) про работу хуков на английском языке.
> * 💡 Для вдохновения можно посмотреть [модули, сделаные компанией Flant](existing_modules/modules.md).
> 
> ## Вопросы
> 
> На случай если у вас остались вопросы после прочтения документации, основным способом для связи является Slack компании Флант, канал `#tech-deckhouse-modules`. Там вы можете задать свой вопрос команде разработки Deckhouse.

Reviewer 2: ```suggestion
Чтобы понять принципы, по которым работают модули Deckhouse, ознакомьтесь с [addon-operator](https://github.com/flant/addon-operator) и [shell-operator](https://github.com/flant/addon-operator).
	
	* 📚 Прочитайте документацию операторов о концепции хуков, например, [что такое конфигурация хука и какие функции она предоставляет](https://flant.github.io/shell-operator/HOOKS.html#hook-configuration). При помощи конфигурации настраиваются данные, которые будут доступны из хука.
	* Ознакомьтесь с информацией о [биндингах](https://flant.github.io/addon-operator/HOOKS.html#bindings). Биндинги - это события, при которых срабатывает хук. Биндинги указываются в конфигурации хука. Хук может сработать не только из-за событий в Kubernetes, а также, например, по расписанию или перед запуском модуля.
	> Хук позволяет сохранять значения в памяти и использовать их позже при рендеринге шаблонов Helm. Об этой особенности и цикле работы модуля, рекомендуем прочитать в [документации Hooks and Helm values](https://flant.github.io/addon-operator/OVERVIEW.html#hooks-and-helm-values).
	* Ознакомьтесь с [концепцией снепшотов](https://flant.github.io/shell-operator/HOOKS.html#snapshots). С помощью снепшотов можно игнорировать отдельные события и реализовать шаблон reconciliation loop, при котором состояние из снепшота приводится к состоянию модуля.
	 > Этим способом в Deckhouse Kubernetes Platform реализовано 100% поддержка хуков во внутренних модулях.
	 * Кроме того, хуки можно использовать вместо `prometheus exporter`. Хуки могут предоставлять метрики, которые Deckhouse будет экспортировать. Ознакомьтесь с информацией [о метриках](https://flant.github.io/addon-operator/metrics/METRICS_FROM_HOOKS.html#custom-metrics).
	* 🎬 В [видео 2019](https://www.youtube.com/watch?v=1_55KPHjVTU) года @andrey.polovov подробно рассказал о том, что такое shell-operator и addon-operator.
	* 🎬 [Более подробное видео](https://www.youtube.com/watch?v=we0s4ETUBLc) про работу хуков на английском языке.
	* 💡 В качестве примера можно посмотреть на [модули, разработанные компанией Flant](existing_modules/modules.md).

## Вопросы

Если после прочтения документации у вас останутся вопросы, то для связи с командой разработки Deckhouse Kubernetes Platform лучше всего использовать Slack компании Флант. В канале `#tech-deckhouse-modules` вы сможете оставить свой вопрос, и команда поможет вам разобраться.
```

---
title: "Модули Deckhouse Kubernetes Platform"
permalink: ru/modules-docs/
lang: ru
---

<p align="center">
<img src="../images/modules-docs/modules_logo.png" alt="deckhouse modules logo" />
</p>

В этом разделе рассказано, как создать собственный модуль Deckhouse Kubernetes Platform (DKP). Существует три основных этапа жизни модуля:

> * **Разработка**: создание кода модуля и его структуры в git-репозитории. В разделе **Анатомия модуля** --> **Папка с модулем** рзбирается, какие папки существуют и что в них находится.

Reviewer 2: или - ...описывается...

> * **Разработка**: создание кода модуля и его структуры в git-репозитории. В разделе **Анатомия модуля** --> **Папка с модулем** рзбирается, какие папки существуют и что в них находится.

Author: Поправь кка считаешь нужным. Я очепятку подсветил просто

> * **Разработка**: создание кода модуля и его структуры в git-репозитории. В разделе **Анатомия модуля** --> **Папка с модулем** рзбирается, какие папки существуют и что в них находится.

Reviewer: норм, резолвлю

> * **Сборка**: создание артефакта модуля во время сборки и загрузка его в registry контейнера. В разделе **Анатомия модуля** --> **Registry контейнера** разбирается, где хранятся образы в registry и какие образы доступны.

Reviewer 2: container registry - после обсуждения глоссария - да.

> * **Разработка**: создание кода модуля и его структуры в git-репозитории. В разделе **Анатомия модуля** --> **Папка с модулем** рзбирается, какие папки существуют и что в них находится.
> * **Сборка**: во время сборки создается артефакт модуля и загружается в registry контейнера. В разделе **Анатомия модуля** --> **Registry контейнера** разбирается, где хранятся образы в registry и какие образы доступны.
> * **Запуск в кластере**: как модуль устанавливается в кластер Deckhouse Kubernetes Platform, как его включить, настроить и проверить, что он работает правильно. В разделе **Как опубликовать модуль в кластере DKP?** разбираются все доступные кастомные ресурсы _Definition_ и содержатся подсказки, где искать ошибки в случае проблем.

Reviewer 2: ```suggestion
* **Разработка**: создание кода модуля и его структуры в git-репозитории. В разделе **Анатомия модуля** --> **Папка с модулем** рзбирается, какие папки существуют и что в них находится.
* **Сборка**: создание артефакта модуля во время сборки и загрузка его в registry контейнера. В разделе **Анатомия модуля** --> **Registry контейнера** разбирается, где хранятся образы в registry и какие образы доступны.
* **Запуск в кластере**: установка модуля в кластер Deckhouse Kubernetes Platform, его включение, настройка и проверка, что он работает правильно. В разделе **Как опубликовать модуль в кластере DKP?** разбираются все доступные кастомные ресурсы _Definition_ и содержатся подсказки, где искать ошибки в случае проблем.
```

## Полезно перед чтением

Чтобы понять принципы, по которым работают модули Deckhouse, ознакомьтесь с [addon-operator](https://github.com/flant/addon-operator) и [shell-operator](https://github.com/flant/addon-operator).

* 📚 Прочитайте документацию операторов о концепции хуков, например, [что такое конфигурация хука и какие функции она предоставляет](https://flant.github.io/shell-operator/HOOKS.html#hook-configuration). При помощи конфигурации настраиваются данные, которые будут доступны из хука.
* Ознакомьтесь с информацией о [биндингах](https://flant.github.io/addon-operator/HOOKS.html#bindings). Биндинги - это события, при которых срабатывает хук. Биндинги указываются в конфигурации хука. Хук может сработать не только из-за событий в Kubernetes, а также, например, по расписанию или перед запуском модуля.
> Хук позволяет сохранять значения в памяти и использовать их позже при рендеринге шаблонов Helm. Об этой особенности и цикле работы модуля, рекомендуем прочитать в [документации Hooks and Helm values](https://flant.github.io/addon-operator/OVERVIEW.html#hooks-and-helm-values).

> * Ознакомьтесь с [концепцией снепшотов](https://flant.github.io/shell-operator/HOOKS.html#snapshots). С помощью снепшотов можно игнорировать отдельные события и реализовать шаблон reconciliation loop, при котором состояние из снепшота приводится к состоянию модуля.

Reviewer 2: За снэпшоты))

> * Ознакомьтесь с [концепцией снепшотов](https://flant.github.io/shell-operator/HOOKS.html#snapshots). С помощью снепшотов можно игнорировать отдельные события и реализовать шаблон reconciliation loop, при котором состояние из снепшота приводится к состоянию модуля.

Reviewer: резолвлю, этот тред, чтобы не плодить треды. учла замечания в сажесте. предлагаю обсуждать под ним

> * Ознакомьтесь с [концепцией снепшотов](https://flant.github.io/shell-operator/HOOKS.html#snapshots). С помощью снепшотов можно игнорировать отдельные события и реализовать шаблон reconciliation loop, при котором состояние из снепшота приводится к состоянию модуля.

Reviewer: ```suggestion
* Ознакомьтесь с [концепцией снэпшотов](https://flant.github.io/shell-operator/HOOKS.html#snapshots). С помощью снэпшотов можно реализовать шаблон reconciliation loop, при котором состояние из снэпшота приводится к состоянию модуля.
```

Reviewer: @z9r5 что думаешь о таком варианте?

Reviewer 2: имхо, норм. вариант. Кэш, хэш, снэпшот...

>  > Этим способом в Deckhouse Kubernetes Platform реализована 100% поддержка хуков во внутренних модулях.

Author: ```suggestion
 > Этим способом в Deckhouse Kubernetes Platform реализована 100% поддержка хуков во внутренних модулях.
```
Ну смысл же был другой в изначальной формулировке - "реализовано 100% хуков".  А превратилось в 100%-тную поддержку.   Как вариант можно что-то типа - "реализована поддержка 100% (всех?) хуков"

Reviewer 2: Да, реализована поддержка всех существующих хуков, например.

Author: Нет. Было "Этим способом в Deckhouse реализовано 100% хуков". Другими словами - все хуки в Deckhouse написаны "на снэпшотах", а не на подписке на события.  А получилась какая-то поддержка хуков не понятная - вообще другое.

Reviewer 2: Перепишем)

Reviewer: ок, решили, резал

* Кроме того, хуки можно использовать вместо `prometheus exporter`. Хуки могут предоставлять метрики, которые Deckhouse будет экспортировать. Ознакомьтесь с информацией [о метриках](https://flant.github.io/addon-operator/metrics/METRICS_FROM_HOOKS.html#custom-metrics).

> * 🎬 В [видео 2019](https://www.youtube.com/watch?v=1_55KPHjVTU) года Андрей Половов подробно рассказал о том, что такое shell-operator и addon-operator.

Author: ```suggestion
* 🎬 В [видео 2019](https://www.youtube.com/watch?v=1_55KPHjVTU) года  подробно рассказал о том, что такое shell-operator и addon-operator.
```

Тут бы пеерформулировать в что-т типа "смотри видео о том, что такое shell-operator и addon-operator". Ссылку оставить только для русской доки. А английское видео толькол для английской

> * 🎬 [Более подробное видео](https://www.youtube.com/watch?v=we0s4ETUBLc) про работу хуков на английском языке.

Author: Убрать в русской, оставить в английской

Author: Давайте оставим комментарий, чтобы не забыть

> * 🎬 [Более подробное видео](https://www.youtube.com/watch?v=we0s4ETUBLc) про работу хуков на английском языке.

Reviewer 2: ```suggestion
```

> * 💡 В качестве примера можно посмотреть на [модули, разработанные компанией Flant](existing_modules/modules.md).

Author: ```suggestion
* 💡 В качестве примера можно посмотреть на [уже разработанные модули](@TODO Подумать куда сослаться).
```

Reviewer 2: Раз речь о модулях, наверное, нужно сослаться на конкретную доку модулей? Либо на доку по ДКП: https://deckhouse.ru/documentation/v1.58/deckhouse-overview.html

Author: Тут не понятно куда ссылаться. Уточним позже.

## Вопросы по документации

> Если после прочтения документации у вас останутся вопросы, то для связи с командой разработки Deckhouse Kubernetes Platform лучше всего использовать Slack компании Флант. В канале `#tech-deckhouse-modules` вы сможете оставить свой вопрос, и команда поможет вам разобраться.

Author: ```suggestion
Присоединяйтесь к [сообществу](/community/), где вам обязательно помогут ответить на появившиеся вопросы.
```

Reviewer: ок, резолв
